### PR TITLE
Remove invalid canonicalization rule 3.4: hoisting unions out of array items

### PR DESCRIPTION
**Pull request by trevorr**
Transforming `(X|Y)[]` into `X[] | Y[]` is invalid. With that step removed,
the remainder of rule 3 can be greatly simplified.

This supersedes #89.

### CODE REVIEW

**Review by ForsakenHarmony — APPROVED** [review]
LGTM

**Comment by jstoiko** [comment]
waiting for @antoniogarrote ‘a green light to merge this.

**Review by antoniogarrote — APPROVED** [review]
This is an error in the algorithm, we need to change it in AMF as well.

**Review by jstoiko — APPROVED** [review]

**Pull request merged**
